cmd/util/ledger/migrations/cadence_values_migration_test.go

```diff
@@ -0,0 +1,232 @@
+			},
+			cadenceValueMigrationReportEntry{
+				Address: interpreter.AddressPath{
+					Address: address,
+					Path: interpreter.PathValue{
+						Identifier: "string_value_2",
+						Domain:     common.PathDomainStorage,
+					},
+				},
+				Migration: "StringNormalizingMigration",
+			},
+			cadenceValueMigrationReportEntry{
+				Address: interpreter.AddressPath{
+					Address: address,
+					Path: interpreter.PathValue{
+						Identifier: "type_value_1",
+						Domain:     common.PathDomainStorage,
+					},
+				},
+				Migration: "AccountTypeMigration",
+			},
+		},
+	)
+}
+
+func payloadsFromEmulatorSnapshot(t *testing.T, snapshotPath string) []*ledger.Payload {
+	db, err := sql.Open("sqlite", snapshotPath)
+	require.NoError(t, err)
+
+	rows, err := db.Query("SELECT key, value FROM ledger")
+	require.NoError(t, err)
+
+	var payloads []*ledger.Payload
+
+	for rows.Next() {
+		var hexKey, hexValue string
+
+		err := rows.Scan(&hexKey, &hexValue)
+		require.NoError(t, err)
+
+		key, err := hex.DecodeString(hexKey)
+		require.NoError(t, err)
+
+		value, err := hex.DecodeString(hexValue)
+		require.NoError(t, err)
+
+		registerId := registerIDKeyFromString(string(key))
+
+		// Type loading currently fails, because the core-contracts
+		// in the emulator snapshot are not migrated yet.
+		// So skip the values that get stored by default, and
+		// keep only the explicitly stored values in 'storage' domain.
+		if registerId.Key == "public" || registerId.Key == "private" {
+			continue
+		}
+
+		ledgerKey := convert.RegisterIDToLedgerKey(registerId)
+
+		payloads = append(
+			payloads,
+			ledger.NewPayload(
+				ledgerKey,
+				value,
+			),
+		)
+	}
+
+	return payloads
+}
+
+// registerIDKeyFromString is the inverse of `flow.RegisterID.String()` method
+func registerIDKeyFromString(s string) flow.RegisterID {
+	parts := strings.SplitN(s, "/", 2)
+
+	owner := parts[0]
+	key := parts[1]
+
+	address, err := common.HexToAddress(owner)
+	if err != nil {
+		panic(err)
+	}
+
+	var decodedKey string
+
+	switch key[0] {
+	case '$':
+		b := make([]byte, 9)
+		b[0] = '$'
+
+		int64Value, err := strconv.ParseInt(key[1:], 10, 64)
+		if err != nil {
+			panic(err)
+		}
+
+		binary.BigEndian.PutUint64(b[1:], uint64(int64Value))
+
+		decodedKey = string(b)
+	case '#':
+		decoded, err := hex.DecodeString(key[1:])
+		if err != nil {
+			panic(err)
+		}
+		decodedKey = string(decoded)
+	default:
+		panic("Invalid register key")
+	}
+
+	return flow.RegisterID{
+		Owner: string(address.Bytes()),
+		Key:   decodedKey,
+	}
+}
+
+type testReportWriterFactory struct{}
+
+func (_m *testReportWriterFactory) ReportWriter(dataNamespace string) reporters.ReportWriter {
+	return &testReportWriter{}
+}
+
+type testReportWriter struct {
+	entries []any
+}
+
+var _ reporters.ReportWriter = &testReportWriter{}
+
+func (r *testReportWriter) Write(entry any) {
+	r.entries = append(r.entries, entry)
+}
+
+func (r *testReportWriter) Close() {}
```

Comment:
Maybe this can be moved to the report writers.

Reply:
This is only used for testing. So I added it here in the test file itself.
Maybe we can move to a common place if other tests also need a similar one.